frontend/src/react/components/appList.jsx

```diff
@@ -187,20 +208,197 @@ class App extends React.Component {
       // copy text
       document.execCommand('copy');
       this.setState({
-        copied: true
+        tokenCopied: true
       });
       tokenElement.blur();
     }catch (err) {
       alert('please press Ctrl/Cmd+C to copy');
     }
   }
 
-  setCopyText(){
+  copyClientId(e){
+    e.preventDefault();
+
+    let clientIdElement = this.refs.clientId;
+    clientIdElement.select();
+
+    try {
+      // copy text
+      document.execCommand('copy');
+      this.setState({
+        clientIdCopied: true
+      });
+      clientIdElement.blur();
+    }catch (err) {
+      alert('please press Ctrl/Cmd+C to copy');
+    }
+  }
+
+  copyClientSecret(e){
+    e.preventDefault();
+
+    let clientSecretElement = this.refs.clientSecret;
+    clientSecretElement.select();
+
+    try {
+      // copy text
+      document.execCommand('copy');
+      this.setState({
+        clientSecretCopied: true
+      });
+      clientSecretElement.blur();
+    }catch (err) {
+      alert('please press Ctrl/Cmd+C to copy');
+    }
+  }
+
+  saveCallbackUrl(e){
+    e.preventDefault();
+
+    let that = this;
+
+    fetch('/dashboard/api/setcallbackurl/', {
+      method: 'POST',
+      credentials: 'include',
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+        'X-CSRFToken': Cookies.get('csrftoken')
+      },
+      body: 'app_id=' + this.props.appId + '&callback_url=' + encodeURIComponent(this.state.callbackUrl)
+    }).then((res)=>{
+      if(res.ok){
+        return res.json();
+      }else{
+        throw new Error('An error occured');
+      }
+    }).then((json)=>{
+      if(json.success){
+        that.setState({
+          callbackUrlSaved: true
+        })
+      }else{
+        throw new Error(json.message);
+      }
+    }).catch((err)=>{
+      that.setState({
+        error: err.message
+      });
+    });
+  }
+
+  setCopyTextToken(){
+    this.setState({
+      tokenCopied: false
+    });
+  }
+
+  setCopyTextClientId(){
     this.setState({
-      copied: false
+      clientIdCopied: false
+    });
+  }
+
+  setCopyTextClientSecret(){
+    this.setState({
+      clientSecret: false
+    });
+  }
+
+  setSaveCallbackUrl() {
+    this.setState({
+      callbackUrlSaved: false
+    })
+  }
+
+  updateRoomBookingsScope(e){
```

Comment:
as mentioned before, we can probably be a lot more DRY by merging these separate endpoints into one.